### ISSUE TITLE
Proposal for Search Improvement

### ISSUE THREAD

**Issue by carlisia** [golang/gddo#316]
## Motivation

Although the Godoc.org site does support a search feature, it performs only a "contains" match against the set of cached import paths. While useful, it only aids users who have a good idea of what they are looking for. The search improvement in thhis proposal would allow better keyword matching against all project-related metadata and documentation text.

## Improvement: Search by new types of matching

We propose that the search feature be extended to support new types of matching and to expand the corpus text against which matches are performed.

- Include the full cached documentation text in the search context. 

- Tokenize the search phrase and match using an implied "AND" operator. For example, "mysql driver" would match "this driver is for mysql" and "go-sql-driver/mysql"

- Stemming: example ‘implement’ would also return ‘implementing’

- Fuzzy search (approximate match): example ‘packge’ returns ‘package’

### Contributors to this proposal
@carlisia
@gdey

**Comment by garyburd** [comment]
Stemming is implemented, but may not work as expected in all cases.  The 'implement' example is one such case. Please file a separate issue for fixing the stemmer.

Search phrases are tokenized and matched using an implied AND.  Please file a separate issue for this with examples of where the query is not working as expected.

**Comment by rafaeljusto** [comment]
@garyburd You are right about the current search already using an implied search. We also tough it would be great to extend this search also to the documentation, instead of only on the package name and description. This could give us more context and increase the result matches. What do you think about it?

**Comment by garyburd** [comment]
It's easy to experiment with the current search code to see if indexing more of the document improves search results. You can find the indexing code in [database/index.go](https://github.com/golang/gddo/blob/master/database/index.go). The [gddo-admin reindex](https://github.com/golang/gddo/blob/master/gddo-admin/reindex.go) subcommand updates the database for any changes to the code. Ask Andrew for a dump of the current database for testing. The gddo-server & Redis commands run comfortably on an inexpensive laptop.

I expect it will be difficult to add more advanced search features to the current search code. It may make sense to throw out the current code and use an existing search server or library. 

Experimenting with other search implementations outside the context of the gddo-server is easy.  Use the [database Do](http://godoc.org/github.com/golang/gddo/database#Database.Do) method to copy data from gddo to another server/library. Write a simple app to test queries.

Before spending a significant time amount of time on this, I suggest running your ideas past Andrew.

**Comment by adg** [comment]
I can provide a dump of the database. Mail me at adg@golang.org.

> Include the full cached documentation text in the search context.

This seems overdue, and we should do it. It's going to require some experimentation to get it right, though.

> Fuzzy search (approximate match): example ‘packge’ returns ‘package’

I think this is of (significantly) lower priority.

--

On the topic of search quality: if we're going to start hacking a lot on making search better, we should have some tests that indicate how much our changes affect search quality. Something to think about.

**Comment by gdey** [comment]
Yes, having test to show progress is good. I think we should add this is the first thing that gets done. But how do you measure such quality? What does better search results empirically mean?

**Comment by adg** [comment]
@gdey that's the challenge ;-)
